File: mvvm-viewbinding/src/main/java/com/everdeng/android/lib/mvvm/viewbinding/base/BaseMvvmFragment.java
package com.everdeng.android.lib.mvvm.viewbinding.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewbinding.ViewBinding;

import com.everdeng.android.lib.core.base.AbstractFragment;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * @author allever
 */
public abstract class BaseMvvmFragment<VB extends ViewBinding, VM extends ViewModel> extends AbstractFragment {

    protected VB mBinding;
    protected VM mViewModel;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Class modelClass;
        Type type = getClass().getGenericSuperclass();
        if (type instanceof ParameterizedType) {
            modelClass = (Class) ((ParameterizedType) type).getActualTypeArguments()[1];
        } else {
            modelClass = BaseViewModel.class;
        }

        mViewModel = (VM) new ViewModelProvider(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                try {
                    return modelClass.getDeclaredConstructor().newInstance();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                } catch (java.lang.InstantiationException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                } catch (NoSuchMethodException e) {
                    e.printStackTrace();
                }

                return null;
            }
        }).get(modelClass);

        observeData();

    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        mBinding = bindView(inflater);
        return mBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mBinding = null;
    }

    protected abstract VB bindView(LayoutInflater inflater);

    protected abstract void observeData();

}

File: mvvm/src/main/java/com/allever/android/lib/mvvm/base/BaseMvvmFragment.kt
package com.allever.android.lib.mvvm.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.allever.android.lib.core.base.AbstractFragment
import java.lang.reflect.ParameterizedType

abstract class BaseMvvmFragment<VM : ViewModel, DB : ViewDataBinding> : AbstractFragment() {

    protected lateinit var mBinding: DB
    protected lateinit var mViewModel: VM

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false)
        mBinding.lifecycleOwner = this
        mViewModel =
            ViewModelProvider(this, object : ViewModelProvider.Factory {
                override fun <VM : ViewModel?> create(modelClass: Class<VM>): VM {
                    val clz = (activity?.javaClass?.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<VM>
                    return clz.getDeclaredConstructor().newInstance()
                }
            }).get((this.javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<VM>)
        mBinding.setVariable(getBindingVariable(), mViewModel)
        (mViewModel as BaseViewModel).bind(requireContext(), this)
        initView()
        (mViewModel as BaseViewModel).init()
        return mBinding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        (mViewModel as BaseViewModel).destroy()
    }

    abstract fun getLayoutId(): Int
    abstract fun getBindingVariable(): Int
    abstract fun initView()

}